### ISSUE TITLE
Feature: Autocompletion-support for fish

### ISSUE THREAD

**Issue by Drezil** [pcapriotti/optparse-applicative#169]
I recently switched to fish and noticed that there is no autocompletion.

As the autocompletion for bash is generated i sat down and modified the script generated by optparse-applicative to work for fish.
I did this with "stack" as my testing app.

Bash-Version:

```
_stack()
{
    local cmdline
    CMDLINE=(--bash-completion-index $COMP_CWORD)

    for arg in ${COMP_WORDS[@]}; do
        CMDLINE=(${CMDLINE[@]} --bash-completion-word $arg)
    done

    COMPREPLY=( $(stack "${CMDLINE[@]}") )
}

complete -o filenames -F _stack stack
```

Fish-Version:

```
function _stack
    set cl (commandline)
    set cn (count $cl)
    set tmpline --bash-completion-index $cn
    for arg in cl
      set tmpline $tmpline --bash-completion-word $arg
    end
    for opt in (stack $tmpline)
      echo -e "$opt"
    end
    return 1
end

complete -f -c stack -a '(_stack)'
```

For generalization you just have to replace "stack" on line 1, 8 and 14 with the progname similar to the bash-script defined in https://github.com/pcapriotti/optparse-applicative/blob/master/Options/Applicative/BashCompletion.hs#L76-L91 .
No further additions need to be done.
It would be nice to generate the script via `program --fish-completion-script`.

Fish also supports description of commands (passing a `-d '(_stack_description)'` to the call in the last line) to be displayed along the completion. I don't know if there is similar demand for bash/zsh in that regard, but it would be nice if such support could be added with a similar call (e.g. `program --completion-index 1 --description-word --help`).

**Comment by HuwCampbell** [comment]
Thanks. I was looking into getting zsh done as well, so this seems like an ok thing to tackle at the same time.

**Comment by HuwCampbell** [comment]
I'm currently working on implementing zsh and fish command line completion, I think we can do better than what we have for bash.

Using git as an example for fish and zsh, I think we can match these with no code for the programmer

```fish
add                                     (Add file contents to the index)  log                                                      (Show commit logs)
annex                                                   (Custom command)  merge                     (Join two or more development histories together)
annex-shell                                             (Custom command)  mergetool  (Run merge conflict resolution tools to resolve merge conflicts)
apply             (Apply a patch on a git index file and a working tree)  mv                       (Move or rename a file, a directory, or a symlink)
```

```zsh
add       -- add file contents to the index
bisect    -- find by binary search the change that introduced a bug
branch    -- list, create, or delete branches
checkout  -- checkout a branch or paths to the working tree
clone     -- clone a repository into a new directory
commit    -- record changes to the repository
```


The implementation idea is to add a `--bash-completion-extended`  flag to the parser, and if it's received, also output descriptions if they exist (and maybe an indicator to say we're want _description instead of compadd )

**Comment by HuwCampbell** [comment]
I've adapted your script and fixed a few bugs in it in #255 
The descriptions for fish was tough, but I think it's still possible with zsh.

**Comment by HuwCampbell** [comment]
Thanks for your help!
